# Final EDA Summary Template

This notebook is a short shared template for final EDA summaries. Each EDA section has only three parts: research question, features/methods, and main findings.

Use the shared `df` loaded below when possible. If a section needs a different CSV, copy the loading cell inside that section and change the path there.

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="whitegrid")
pd.set_option("display.max_columns", 100)

PROJECT_ROOT = Path.cwd()
while PROJECT_ROOT.name != "summer26-pokemon-battle-predictor" and PROJECT_ROOT.parent != PROJECT_ROOT:
    PROJECT_ROOT = PROJECT_ROOT.parent

DATA_DIR = PROJECT_ROOT / "data"

# Tomorrow, change only this path if everyone is using the same parsed CSV.
PARSED_CSV_PATH = DATA_DIR / "data_cleaned.csv.zip"

print("Project root:", PROJECT_ROOT)
print("Parsed CSV path:", PARSED_CSV_PATH)

Project root: /Users/gknapp/Library/CloudStorage/GoogleDrive-gknapp@colby.edu/My Drive/Technology/Coding/Personal Notes/Erdos Institute/project/summer26-pokemon-battle-predictor
Parsed CSV path: /Users/gknapp/Library/CloudStorage/GoogleDrive-gknapp@colby.edu/My Drive/Technology/Coding/Personal Notes/Erdos Institute/project/summer26-pokemon-battle-predictor/data/data_cleaned.csv.zip


In [2]:
# Load the shared dataframe.
# Uncomment after PARSED_CSV_PATH points to the correct file.

df = pd.read_csv(PARSED_CSV_PATH)
print(df.shape)
df.head()

(12778, 299)


,format,id,p1_win,ratedQ,n_turns,start_time,end_time,duration,p1name,p1side,p1elo0,p1elo1,p2name,p2side,p2elo0,p2elo1,type_diversity_diff,num_boosting_abilities_diff,num_move_boosters_diff,total_stat_diff,p1_total_adv,p1_revealed_team_size,p2_revealed_team_size,M11_name,M11_speciesId,M11_used,M11_gender,M11_shinyQ,M11_level,M11_ability,M11_item,M11_teraType,M11_role,M11_mv1,M11_mv2,M11_mv3,M11_mv4,M11_type1,M11_type2,M11_hp,M11_atk,M11_def,M11_spa,M11_spd,M11_spe,M11_off,M12_name,M12_speciesId,M12_used,M12_gender,...,M24_spa,M24_spd,M24_spe,M24_off,M25_name,M25_speciesId,M25_used,M25_gender,M25_shinyQ,M25_level,M25_ability,M25_item,M25_teraType,M25_role,M25_mv1,M25_mv2,M25_mv3,M25_mv4,M25_type1,M25_type2,M25_hp,M25_atk,M25_def,M25_spa,M25_spd,M25_spe,M25_off,M26_name,M26_speciesId,M26_used,M26_gender,M26_shinyQ,M26_level,M26_ability,M26_item,M26_teraType,M26_role,M26_mv1,M26_mv2,M26_mv3,M26_mv4,M26_type1,M26_type2,M26_hp,M26_atk,M26_def,M26_spa,M26_spd,M26_spe,M26_off
0,gen9randombattle,2642076539,0,True,21,1782884765,1782884937,172,Lirnom,1,1624,1607,Jenome,2,1674,1691,0,0,1,-79,3.125322,4,5,Cyclizar,cyclizar,1,M,False,83,Regenerator,Heavy-Duty Boots,Dragon,Fast Support,taunt,rapidspin,shedtail,dracometeor,Dragon,Normal,252,205,156,189,156,249,205,Cramorant,cramorant,1,M,...,182,118,158,198,Tsareena,tsareena,1,F,False,87,Queenly Majesty,Life Orb,Steel,Fast Support,tripleaxel,synthesis,highjumpkick,powerwhip,Grass,NaN,267,259,220,137,220,175,259,Toxtricity,toxtricity,1,F,False,82,Punk Rock,Choice Specs,Normal,Fast Attacker,sludgewave,boomburst,voltswitch,overdrive,Electric,Poison,257,165,162,234,162,170,234
1,gen9randombattle,2642074504,1,True,7,1782884343,1782884385,42,jnpr010334,1,1651,1667,masajji,2,1582,1566,-1,-1,-2,19,0.997149,2,3,Virizion,virizion,1,N,False,82,Justified,Life Orb,Rock,Setup Sweeper,stoneedge,closecombat,swordsdance,leafblade,Grass,Fighting,283,195,165,195,259,224,195,Koraidon,koraidon,0,N,...,199,141,141,250,Gouging Fire,gougingfire,0,N,False,74,Protosynthesis,Heavy-Duty Boots,Fairy,Bulky Setup,heatcrash,morningsun,outrage,dragondance,Fire,Dragon,277,213,222,139,181,178,213,Zangoose,zangoose,1,M,False,85,Toxic Boost,Toxic Orb,Normal,Fast Attacker,knockoff,facade,swordsdance,closecombat,Normal,NaN,263,244,151,151,151,202,244
2,gen9randombattle,2641921866,1,False,31,1782859109,1782859432,323,realgreenz,1,0,0,ABIDTHEBEST,2,0,0,0,0,1,129,5.479100,6,6,Plusle,plusle,1,M,False,95,Lightning Rod,Life Orb,Fairy,Setup Sweeper,thunderbolt,nastyplot,encore,alluringvoice,Electric,NaN,268,100,130,215,196,234,215,Hatterene,hatterene,1,F,...,188,236,219,188,Passimian,passimian,1,M,False,83,Defiant,Leftovers,Steel,Bulky Setup,gunkshot,bulkup,drainpunch,knockoff,Fighting,NaN,302,247,197,114,147,180,247,Tornadus,tornadus,1,M,False,82,Prankster,Heavy-Duty Boots,Flying,Fast Attacker,nastyplot,bleakwindstorm,grassknot,focusblast,Flying,NaN,264,193,162,252,178,229,252
3,gen9randombattle,2641831207,1,True,51,1782847154,1782847410,256,soumatou_story,1,2113,2127,TheLastOne,2,2014,2000,-2,-2,2,347,3.903277,6,6,Tsareena,tsareena,1,F,False,87,Queenly Majesty,Choice Scarf,Steel,Fast Support,tripleaxel,knockoff,powerwhip,uturn,Grass,NaN,267,259,220,137,220,175,259,Snorlax,snorlax,1,F,...,143,179,270,234,Hitmonchan,hitmonchan,1,M,False,87,Iron Fist,Leftovers,Steel,Bulky Support,encore,drainpunch,knockoff,poisonjab,Fighting,NaN,229,232,187,111,241,182,232,Corviknight,corviknight,1,M,False,80,Mirror Armor,Rocky Helmet,Dragon,Bulky Support,uturn,bodypress,bravebird,roost,Flying,Steel,288,185,214,131,182,153,185
4,gen9randombattle,2642210572,1,True,18,1782911409,1782911669,260,jobzy,1,1177,1201,danbur,2,1165,1141,0,2,1,-120,1.475808,2,6,Dewgong,dewgong,1,M,False,94,Thick Fat,Assault Vest,Poison,Bulky Attacker,knockoff,tripleaxel,surf,flipturn,Water,Ice,322,185,204,185,232,185,185,Iron Boulder,ironboulder,0,N,...,152,170,195,256,Jirachi,jirachi,1,N,False,80,Serene Grace,Leftovers,Water,Fast Support,bodyslam,ironhead,protect,wish,Steel,Psychi

See `data/README.md` for a thorough description of the column headers.

## 1. General Pokemon EDA

### Research Question

-

### Features Looking Into And Methods

Focus: General Pokemon stats, types, and team composition.

-

### Main Findings

-

In [3]:
# Optional: create the key summary table or plot for this section.
# Use the shared df, or load a section-specific CSV here if needed.

# summary_table = ...
# summary_table

## 2. First Pokemon Advantage EDA

### Research Question

-

### Features Looking Into And Methods

Focus: Opening Pokemon / first matchup advantage.

-

### Main Findings

-

In [4]:
# Optional: create the key summary table or plot for this section.
# Use the shared df, or load a section-specific CSV here if needed.

# summary_table = ...
# summary_table

## 3. Active Switch Count And Win Rate EDA

### Research Question

-

### Features Looking Into And Methods

Focus: Relationship between active switches during battle and win rate.

-

### Main Findings

-

In [5]:
# Optional: create the key summary table or plot for this section.
# Use the shared df, or load a section-specific CSV here if needed.

# summary_table = ...
# summary_table

## 4. Pokemon Advantage Stats EDA

### Research Question

- Can we teach the model about individual pokemon interactions (e.g. 'water beats fire' and 'a large defense stat is good against physical attackers but useless against special attackers') without requiring it to learn those interactions from team-level data (e.g. this team of six pokemon beat this other team of six pokemon)?

This question carries major significance for our project.  After all, an experienced Pokemon player is able to rapidly analyze individual matchups and uses that information to inform their strategies.  Our final model should be able to use the same information the players are using.  The primary subquestion is then:

  - What information are the players using to make decisions during a battle?

#### Example

As an instructive example of the kind of analysis that an experienced player typically does, consider a hypothetical matchup of Tauros-Paldea-Aqua against Delphox.

##### Major factors

<u>Type</u>

Tauros is a Fighting-Water type while Delphox is Fire-Psychic.  If Tauros uses a water move into Delphox, it will be supereffective (i.e. the damage dealt to the Delphox will be multiplied by 2), while if Delphox uses a Psychic move into Tauros, it will also be supereffective.  With only information about the types of the two Pokemon, it's hard to say who has the advantage.

<u>Statistics</u>

Let's next look at the statistics of these two Pokemon.  Neither Pokemon here is the most powerful attacker.  Delphox uses special attacks, and its special attack stat is 240, which is average to below average for a special attacker.  Tauros uses physical attacks, and its attack stat is 225, which is below average for a physical attacker.  Their respective defensive stats are also somewhat low: Tauros' special defense stat is 160, and Delphox's defense stat is 169.  Their HP stats are again, quite similar with the Delphox having 263 hit points and the Tauros having 254 hit points.  Finally, Delphox is the faster of the two Pokemon, so it gets to move first every round.

Given all of this information, it seems like the Delphox has the advantage: the Delphox gets to move first, and the amount of damage each is doing to the other should be fairly comparable.  Without any more information, an experienced player might guess that in a one-on-one matchup between these two Pokemon, the Delpox will knock out the Tauros after two rounds of combat.  This is only a modest advantage however, since there are situations where the Tauros could beat the Delphox.  For instance, if the Tauros begins the one-on-one matchup with full HP and the Delphox enters the matchup with only half of its HP, the Tauros will take lots of damage from the Delphox's attack, but will KO the Delphox in response.

##### Minor Factors

That said, there are other factors at play here, like which items the Pokemon are holding, or which abilities they have.

<u>Abilities</u>

Different Pokemon have different special abilities which can alter matchups.  Neither Pokemon's ability is relevant to this particular matchup.

<u>Items</u>

Delphox sometimes carries the Choice Specs item in this format which boosts its special attack stat to well-above average levels.  Tauros often carries the Choice Band item, which has the analogous effect for its stats.

If the Delphox has Choice Specs, it now dominates the Tauros: it moves first, and KOs the Tauros before the Tauros can do anything in return.  If the Delphox does not have Choice Specs, while the Tauros has a Choice Band, the Tauros now has a small advantage: in the first round of combat, the Tauros takes lots of damage from the Delphox, but manages to KO the Delphox on its turn.

<u>Known Moves</u>

The analysis up to this point has been assuming that Tauros knows a generic water move and Delphox knows a generic psychic move, but what if their moves aren't generic?  If the Tauros knows Wave Crash (a very powerful water move that deals recoil damage to the Tauros), then the matchup may end up with both pokemon being knocked out as the Delphox first deals damage to the Tauros, and the Tauros responds with Wave Crash, knocking out the Delphox, and then knocking out the Tauros with the recoil damage.  Alternatively, if the Tauros knows a less powerful water move without the recoil (say, Liquidation), but the Tauros also knows Aqua Jet (a water move without much power, but which allows the Tauros to go before the Delphox), its possible that the Tauros could knock out Delphox in the second round of combat.

##### Conclusion

If our model trains on a data set that only contains the 12 pokemon involved in a battle, their types, abilities, items, known moves (all represented by strings so far), and stats (represented by ints), the model is unlikely to learn the feature interactions that players use all of the time.  Indeed, it might only ever see one match in which Tauros-Paldea-Aqua is on one team and Delphox is on the other, and in that match, it will only see which team won.  Instead, we need to find some way of telling the model the information that we already know about how these features interact.

### Features And Methods

In order to account for the matchups of individual pokemon, we created the advantage statistic, which measures how big of an advantage one Pokemon has over the other.  Advantage is a damage-based statistic.

Given two Pokemon, say Tauros-Paldea-Aqua and Delphox, the advantage statistic starts by approximating how much damage each Pokemon would deal to the other with its best single generic attack (as a proportion of the defending Pokemon's HP).  For instance, it will calculate how much average damage Delphox will do to Tauros with a generic Psychic attack (let's say 80%) and how much damage Tauros will do to Delphox with a generic Water attack (let's say 70%).

From those numbers, it will "simulate" a one-on-one matchup between those two Pokemon, with each Pokemon assumed to only select its best damaging move.  With the numbers above, on the first turn, Delphox deals 80% damage to Tauros, while Tauros deals 70% damage to Delphox.  On the next turn, Delphox KOs Tauros and Tauros does not get to deal any more damage.

The advantage statistic is the damage differential in given matchup.  In the above example, Delphox deals damage equal to 2*80% = 160% of Tauros' HP, while Tauros deals damage equal to 70% of Delphox's HP.  Then advantage(Delphox,Tauros) = 160% - 70% = 90%, while advantage(Tauros,Delphox) = 70% - 160% = -90%.

In a battle, then, where each team has 6 pokemon, we computed the 36 distinct (up to sign) advantage stats of the form advantage(m1,m2) where m1 is a Pokemon on team 1 and m2 is a Pokemon on team 2.  The total advantage of player 1 is defined to be the sum of those 36 advantage stats.

Note: why train on the 1-dimensional total advantage stat rather than the 36-dimensional list of advantage stats?  The primary reason is that we know that across all battles, each matchup should carry the same weight.  While it might be true for a particular match that player 1 Pokemon 3's advantage over player 2 Pokemon 5 means more than player 1 Pokemon 4's advantage over player 2 Pokemon 1, those two matchups should carry the same weight across all battles.  (It's also worth noting that models which trained on total advantage performed better than those that trained on the 36 individual stats.)

### Main Findings

-

In [6]:
# Optional: create the key summary table or plot for this section.
# Use the shared df, or load a section-specific CSV here if needed.

# summary_table = ...
# summary_table

### Known Limitations

There are a number of limitations to the advantage stat.

First and foremost, the advantage stat uses a damage approximation.  We have the exact formula for damage calculations, but it is incredibly complex to implement due to the vast array of moves, abilities, items, and field conditions.  While our damage calculator accounts for the major factors listed above (types and stats), it handles only some abilities and a few of the most common items.  It does not attempt to account for moves in any way.  The best way to improve upon this would be to learn to access the calculator at calc.pokemonshowdown.com.

Perhaps more subtle is the issue that our advantage stat makes an assumption that each individual Pokemon enters the matchup with full HP.  This tends to punish sweepers/fast wallbreakers like the offensive Deoxys variants who can deal 90% damage to nearly any pokemon, but get knocked out by nearly any attack.  Perhaps we could improve upon this by adjusting the advantage stat so that it depends on current HP rather than maximum HP.  Then, at the beginning of the battle we could take the average of all advantage stats across all HP states (or a substantial selection of them depending on computational complexity).


## 5. Other Traits

# Overall Takeaways

### Research Question

What did the combined EDA work help us understand about Pokemon battle prediction?

-

### Features Looking Into And Methods

-

### Main Findings

-